include/gevolution/lightcone.hpp

```diff
@@ -0,0 +1,287 @@
+            mpi::all_to_all(com,sendrecv_query,sendrecv_query);
+            
+           for(int proc=0;proc<com.size();++proc)
+               for(const auto &key : sendrecv_query[proc])
+               {
+                   sendrecv_result[proc].push_back( {key,local_query(key)} );
+               }
+           
+           mpi::all_to_all(com,sendrecv_result,sendrecv_result);
+            
+        } // clean sendrecv_query
+        
+        std::vector< data_type > result;
+        
+        for(const auto &v : sendrecv_result)
+            for(const auto &d : v)
+                result.push_back(d);
+        
+        return result;
+    }
+};
+
+class parallel_lightcone
+{
+    ::boost::mpi::communicator com;
+    
+    // geometry of the full lightcone
+    // FIXME: initialize this geometry
+    lightcone_geometry geometry
+
+    
+    // in this way we decide where to store the information relative to each particle in the
+    // distributed database.
+    class destination_func
+    {
+        const ::boost::mpi::communicator& c;
+        
+        public:
+        
+        int operator()(const id_type I)
+        {
+            return I % c.size();
+        }
+    };
+    
+    distributed_database<id_type,bool,destination_func> db;
+    
+    // value of the conformal time (cosmic clock) when the lightcone was called
+    double last_tau; 
+    
+    static double sqr(double x){return x*x;}
+    
+    static bool inside_shell(const particle& part, const lightcone_geometry& shell)
+    // returns true if the particle part lies inside the lightcone shell
+    {
+        // comoving position difference
+        const std::array<double,3> pos_diff{
+            part.pos[0]-shell.vertex[0],
+            part.pos[1]-shell.vertex[1],
+            part.pos[2]-shell.vertex[2]
+        };
+        
+        // comoving distance to the observer
+        const double dist = std::sqrt(
+              sqr(pos_diff[0])
+            + sqr(pos_diff[1])
+            + sqr(pos_diff[2])
+            ) ; 
+        
+        // modulo of the direction vector
+        const double dir_mod = std::sqrt( 
+            sqr(shell.direction[0]) +
+            sqr(shell.direction[1]) +
+            sqr(shell.direction[2]) );
+        
+        const double cos_angle =  
+            ( pos_diff[0]*shell.direction[0] + 
+              pos_diff[1]*shell.direction[1] + 
+              pos_diff[2]*shell.direction[2] )
+            / dir_mod / dist;
+        
+        return dist >= shell.distance[0] && dist<shell.distance[1] && cos_angle >= shell.opening;
+    }
+    
+    public:
+    parallel_lightcone(
+          const ::boost::mpi::communicator& in_com, 
+          lightcone_geometry lc, 
+          std::string fname):
+        com{in_com}, geometry{lc}, db(com,destination_func{com}),
+        
+        // initialized to the mininum value possible of the tau (cosmic time) for an observed event
+        last_tau{std::max(0.0,geometry.tau - geometry.distance[1])}
```

Comment:
This will break. Because `lightcone_geometry` has no `tau` but `redshift`. I think we shall make a distinction of data structures. Have a `lightcone_geometry` that records the observer `tau` and the geometrical aspects of the lightcone. And another `lightcone_observation` that deals with it from the observation point of view, so z min, zmax, z observer and so on.